### PR TITLE
[WIP] partially unbreak chunked gzip responses on appengine

### PR DESCRIPTION
**Pull request by t-8ch**
see #618

### CODE REVIEW

**Comment by shazow** [comment]
Added WIP, ping when it's ready for review. :)

File: urllib3/response.py [hunk 1]
```diff
@@ -422,6 +422,18 @@ def _handle_chunk(self, amt):
...
         return returned_chunk
 
+    @staticmethod
+    def _is_response_to_head(response):
+        # FIXME: Can we do this somehow without accessing private httplib _method?
+        if response is None:
+            return False
+
+        method = response._method
+        if isinstance(method, int):  # Platform-specific: Appengine
```
Comment (Lukasa) [inline urllib3/response.py]: Wait, what?

Comment (t-8ch) [inline urllib3/response.py]: Don't blame the messenger :-)

File: urllib3/response.py [hunk 2]
```diff
@@ -437,7 +449,7 @@ def read_chunked(self, amt=None, decode_content=None):
             raise ResponseNotChunked("Response is not chunked. "
                 "Header 'transfer-encoding: chunked' is missing.")
 
-        if self._original_response and self._original_response._method.upper() == 'HEAD':
+        if self._is_response_to_head(self._original_response):
             # Don't bother reading the body of a HEAD request.
             # FIXME: Can we do this somehow without accessing private httplib _method?
```
Comment (shazow) [inline urllib3/response.py]: I think this FIXME is obsolete here, since it got moved into the helper?

File: urllib3/response.py [hunk 3]
```diff
@@ -422,6 +422,18 @@ def _handle_chunk(self, amt):
             self.chunk_left = None
         return returned_chunk
 
+    @staticmethod
```
Comment (shazow) [inline urllib3/response.py]: Why staticmethod? Seems to make sense for it to be a regular instance method and skip passing response in.

File: urllib3/response.py [hunk 4]
```diff
@@ -422,6 +422,18 @@ def _handle_chunk(self, amt):
...
+    def _is_response_to_head(response):
+        # FIXME: Can we do this somehow without accessing private httplib _method?
+        if response is None:
+            return False
+
+        method = response._method
+        if isinstance(method, int):  # Platform-specific: Appengine
+            return method == 3
+        else:
```
Comment (shazow) [inline urllib3/response.py]: Style nit: Can we lose the unnecessary `else:`?

File: urllib3/response.py (hunk 3 above)
Comment (t-8ch) [inline urllib3/response.py]: I thought about moving it to utils/connection or so, where the other platform specific stuff lives.

Comment (shazow) [inline urllib3/response.py]: +0.5 to that. I was thinking the same.

**Comment by agfor** [comment]
I do think this is correct, it just doesn't fix things all the way because the data coming back from GAE httplib is still not chunked.

**Comment by sigmavirus24** [comment]
This needs to be rebased. @t-8ch would you like me to do that?

**Comment by t-8ch** [comment]
Sorry for forgetting this...
@sigmavirus24 This would be awesome, it may take till tomorrow for me to get at my PC

**Comment by sigmavirus24** [comment]
@t-8ch I'll take a stab in a couple hours. :custard:

**Comment by shazow** [comment]
:sparkles: :custard: :sparkles:

**Comment by sigmavirus24** [comment]
Also @t-8ch please don't apologize for forgetting this. You did the hard work of figuring out what int GAE returns for a response to a HEAD request. That would have blocked me.

**Comment by t-8ch** [comment]
@sigmavirus24 @shazow I did the rebase, it was actually trivial :-)

**Comment by shazow** [comment]
Thanks @t-8ch! :custard:

**Pull request merged**